lightning/src/offers/static_invoice.rs

```diff
@@ -0,0 +1,1173 @@
+			}
+		}
+
+		let Self { offer_bytes, invoice, keys } = self;
+		let unsigned_invoice = UnsignedStaticInvoice::new(&offer_bytes, invoice);
+		let invoice = unsigned_invoice
+			.sign(|message: &UnsignedStaticInvoice| {
+				Ok(secp_ctx.sign_schnorr_no_aux_rand(message.tagged_hash.as_digest(), &keys))
+			})
+			.unwrap();
+		Ok(invoice)
+	}
+
+	invoice_builder_methods_common!(self, Self, self.invoice, Self, self, S, StaticInvoice, mut);
+}
+
+/// A semantically valid [`StaticInvoice`] that hasn't been signed.
+pub struct UnsignedStaticInvoice {
+	bytes: Vec<u8>,
+	contents: InvoiceContents,
+	tagged_hash: TaggedHash,
+}
+
+macro_rules! invoice_accessors { ($self: ident, $contents: expr) => {
+	/// The chain that must be used when paying the invoice. [`StaticInvoice`]s currently can only be
+	/// created from offers that support a single chain.
+	pub fn chain(&$self) -> ChainHash {
+		$contents.chain()
+	}
+
+	/// Opaque bytes set by the originating [`Offer::metadata`].
+	///
+	/// [`Offer::metadata`]: crate::offers::offer::Offer::metadata
+	pub fn metadata(&$self) -> Option<&Vec<u8>> {
+		$contents.metadata()
+	}
+
+	/// The minimum amount required for a successful payment of a single item.
+	///
+	/// From [`Offer::amount`].
+	///
+	/// [`Offer::amount`]: crate::offers::offer::Offer::amount
+	pub fn amount(&$self) -> Option<Amount> {
+		$contents.amount()
+	}
+
+	/// Features pertaining to the originating [`Offer`], from [`Offer::offer_features`].
+	///
+	/// [`Offer`]: crate::offers::offer::Offer
+	/// [`Offer::offer_features`]: crate::offers::offer::Offer::offer_features
+	pub fn offer_features(&$self) -> &OfferFeatures {
+		$contents.offer_features()
+	}
+
+	/// A complete description of the purpose of the originating offer, from [`Offer::description`].
+	///
+	/// [`Offer::description`]: crate::offers::offer::Offer::description
+	pub fn description(&$self) -> Option<PrintableString> {
+		$contents.description()
+	}
+
+	/// Duration since the Unix epoch when an invoice should no longer be requested, from
+	/// [`Offer::absolute_expiry`].
+	///
+	/// [`Offer::absolute_expiry`]: crate::offers::offer::Offer::absolute_expiry
+	pub fn absolute_expiry(&$self) -> Option<Duration> {
+		$contents.absolute_expiry()
+	}
+
+	/// The issuer of the offer, from [`Offer::issuer`].
+	///
+	/// [`Offer::issuer`]: crate::offers::offer::Offer::issuer
+	pub fn issuer(&$self) -> Option<PrintableString> {
+		$contents.issuer()
+	}
+
+	/// Paths to the node that may supply the invoice on the recipient's behalf, originating from
+	/// publicly reachable nodes. Taken from [`Offer::paths`].
+	///
+	/// [`Offer::paths`]: crate::offers::offer::Offer::paths
+	pub fn offer_message_paths(&$self) -> &[BlindedPath] {
+		$contents.offer_message_paths()
+	}
+
+	/// Paths to the recipient for indicating that a held HTLC is available to claim when they next
+	/// come online.
+	pub fn message_paths(&$self) -> &[BlindedPath] {
+		$contents.message_paths()
+	}
```

Comment:
It's a bit inconsistent that `Bolt12Invoice::message_paths` and `StaticInvoice::message_paths` don't correspond to the same thing. We may want to re-think the naming here. Further, `message_paths` may not be descriptive enough especially when we also add paths for payment notification to the offer. Though I'm having trouble finding a way to differentiate naming the latter from the one added here.

Possible naming:
- `request_paths` for `offer_paths` TLV
- `payment_paths` for `invoice_paths` and `invoice_blindedpay`
- `notification_paths` for paths add to `offer` for upcoming payment notifications
- ??? for paths added to `invoice` for async payments

WDYT?

Reply:
Don't feel super strongly since users are unlikely to be accessing these fields directly, but I'm happy to rename these methods. Since we're pulling in other considerations like payment notifications and renaming offer fields as well, I'm working on a follow-up based on this PR that should be open sometime today, if that works for you. Naming suggestions SGTM.

Reply:
Opened #3118.